File: frontend/src/lib/errorTracking.ts
import { getAuthToken } from '../api/client'
import type { ErrorTrackingSettings } from '../api/types'

const API_BASE = import.meta.env.VITE_API_BASE_URL ?? '/api/v1'

let config: ErrorTrackingSettings | null = null
let handlersInstalled = false
let configPromise: Promise<ErrorTrackingSettings> | null = null

export async function loadErrorTrackingConfig(): Promise<ErrorTrackingSettings> {
  if (config) return config
  if (configPromise) return configPromise
  configPromise = (async () => {
    try {
      const response = await fetch(`${API_BASE}/errors/config`)
      if (!response.ok) {
        config = { enabled: false, frontend: false, backend: false, api: false }
        return config
      }
      config = (await response.json()) as ErrorTrackingSettings
      return config
    } catch {
      config = { enabled: false, frontend: false, backend: false, api: false }
      return config
    } finally {
      configPromise = null
    }
  })()
  return configPromise
}

export function getErrorTrackingConfig(): ErrorTrackingSettings | null {
  return config
}

export function setErrorTrackingConfig(next: ErrorTrackingSettings): void {
  config = next
}

export function shouldReportFrontendErrors(): boolean {
  return Boolean(config?.enabled && config.frontend)
}

export function reportFrontendError(payload: {
  message: string
  stack_trace?: string | null
  path?: string | null
  method?: string | null
  status_code?: number | null
  page_url?: string | null
  extra?: Record<string, unknown> | null
}): void {
  if (!shouldReportFrontendErrors()) return

  const token = getAuthToken()
  void fetch(`${API_BASE}/errors/report`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify({
      message: payload.message.slice(0, 4096),
      stack_trace: payload.stack_trace?.slice(0, 16384) ?? null,
      path: payload.path?.slice(0, 512) ?? null,
      method: payload.method?.slice(0, 16) ?? null,
      status_code: payload.status_code ?? null,
      page_url: payload.page_url?.slice(0, 1024) ?? window.location.href,
      extra: payload.extra ?? null,
    }),
  }).catch(() => {
    /* never break the app for reporting failures */
  })
}

export function installGlobalErrorHandlers(): void {
  if (handlersInstalled || !shouldReportFrontendErrors()) return
  handlersInstalled = true

  window.addEventListener('error', (event) => {
    if (!shouldReportFrontendErrors()) return
    const message =
      event.message ||
      (event.error instanceof Error ? event.error.message : 'Unhandled error')
    const stack =
      event.error instanceof Error
        ? event.error.stack
        : `${event.filename}:${event.lineno}:${event.colno}`
    reportFrontendError({
      message,
      stack_trace: stack,
      page_url: window.location.href,
      extra: { type: 'window.onerror' },
    })
  })

  window.addEventListener('unhandledrejection', (event) => {
    if (!shouldReportFrontendErrors()) return
    const reason = event.reason
    const message =
      reason instanceof Error
        ? reason.message
        : typeof reason === 'string'
          ? reason
          : 'Unhandled promise rejection'
    const stack = reason instanceof Error ? reason.stack : null
    reportFrontendError({
      message,
      stack_trace: stack,
      page_url: window.location.href,
      extra: { type: 'unhandledrejection' },
    })
  })
}

export async function initErrorTracking(): Promise<void> {
  await loadErrorTrackingConfig()
  installGlobalErrorHandlers()
}

File: backend_api/http/schemas/error_tracking.py
"""Pydantic schemas for error tracking."""

from typing import Any, Literal

from pydantic import BaseModel, Field

ErrorSource = Literal["frontend", "backend", "api"]


class ErrorTrackingSettings(BaseModel):
    enabled: bool = False
    frontend: bool = False
    backend: bool = False
    api: bool = False


class ErrorTrackingSettingsUpdate(BaseModel):
    enabled: bool | None = None
    frontend: bool | None = None
    backend: bool | None = None
    api: bool | None = None


class ErrorReportRequest(BaseModel):
    message: str = Field(..., min_length=1, max_length=4096)
    stack_trace: str | None = Field(default=None, max_length=16384)
    path: str | None = Field(default=None, max_length=512)
    method: str | None = Field(default=None, max_length=16)
    status_code: int | None = None
    page_url: str | None = Field(default=None, max_length=1024)
    extra: dict[str, Any] | None = None


class ErrorEventOut(BaseModel):
    id: int
    source: str
    message: str
    stack_trace: str | None = None
    path: str | None = None
    method: str | None = None
    status_code: int | None = None
    user_id: int | None = None
    user_agent: str | None = None
    page_url: str | None = None
    extra: dict[str, Any] | None = None
    created_at: str | None = None
